src/ptr_template.cpp

```diff
@@ -560,25 +560,28 @@ class createTemplate : public PlayerScript {
         }
     }
 
-    static void AddTemplateHotbar(Player* player, uint32 index) // Someone smarter than me needs to fix this.
+    static void AddTemplateHotbar(Player* player, uint32 index)
     { //                                                    0       1      2
         QueryResult barInfo = WorldDatabase.Query("SELECT Button, Action, Type FROM mod_ptrtemplate_action WHERE (ID = {} AND RaceMask & {} AND ClassMask & {})", index, player->getRaceMask(), player->getClassMask());
-        for (uint8 j = ACTION_BUTTON_BEGIN; j <= MAX_ACTION_BUTTONS; j++) // This is supposed to go through every available action slot and remove what's there.
-        { //                                                                 This doesn't work for spells added by AddTemplateSpells.
-            player->removeActionButton(j); //                                I don't know why and I've tried everything I can think of, but nothing's worked.
-        } //                                                                 And yes, I do want the hotbar cleared for characters that don't have any hotbar data in the template.
         if (barInfo)
         {
+            for (uint8 j = ACTION_BUTTON_BEGIN; j <= MAX_ACTION_BUTTONS; j++)
+            {
+                player->removeActionButton(j);
+            }
             do
             {
                 uint8 buttonEntry = (*barInfo)[0].Get<uint8>();
                 uint32 actionEntry = (*barInfo)[1].Get<uint32>();
                 uint8 typeEntry = (*barInfo)[2].Get<uint8>();
-                player->addActionButton(buttonEntry, actionEntry, typeEntry);
-                LOG_DEBUG("module", "Added hotbar spell {} on button {} with type {} for template character {}.", actionEntry, buttonEntry, typeEntry, player->GetGUID().ToString());
+                if (player->addActionButton(buttonEntry, actionEntry, typeEntry))
+                    LOG_DEBUG("module", "Added hotbar spell {} on button {} with type {} for template character {}.", actionEntry, buttonEntry, typeEntry, player->GetGUID().ToString());
+                else
+                    LOG_ERROR("module", "Failed to add hotbar spell {} on button {} with type {} for template character {}.", actionEntry, buttonEntry, typeEntry, player->GetGUID().ToString());
             } while (barInfo->NextRow());
+            player->SendActionButtons(1);
+            player->SaveToDB(false, false); // commit action buttons
```

Comment:
```suggestion
            player->SaveToDB(false, false); // Commit action buttons
```
Do you happen to know of any reasons this would have to fail?

Reply:
it fails if this check fails https://github.com/sogladev/azerothcore-wotlk/blob/7abbca0f5b4ab3a15d78db7ec40e3b4358bed7da/src/server/game/Entities/Player/Player.cpp#L5566

An example I encountered is when the template is trying to add an action spell that is not learned as the entry is not set in spells template
https://github.com/sogladev/azerothcore-wotlk/blob/7abbca0f5b4ab3a15d78db7ec40e3b4358bed7da/src/server/game/Entities/Player/Player.cpp#L5589C13-L5589C35